File: intrinsic_motivation_ros/src/empowerment_node_planar.cpp
// empowerment_node_planar -- the 2-link version, kept for the simple demo.
//
// Ground-truth outcomes rather than a rendered depth image, which is the
// wrong thing by the paper's definition but is the right place to build
// intuition: it runs in milliseconds and needs no simulator.
//
// Subscribes:
//     joint_states     sensor_msgs/JointState
//     object_pose      geometry_msgs/PoseStamped
//
// Publishes:
//     empowerment/value      std_msgs/Float64
//     empowerment/gradient   geometry_msgs/Vector3
//     empowerment/markers    visualization_msgs/MarkerArray
#include <algorithm>
#include <chrono>
#include <cmath>
#include <memory>
#include <string>
#include <vector>

#include <Eigen/Core>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/vector3.hpp>
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/joint_state.hpp>
#include <std_msgs/msg/float64.hpp>
#include <visualization_msgs/msg/marker_array.hpp>

#include "intrinsic_core/arm2d.hpp"
#include "intrinsic_core/discretise.hpp"
#include "intrinsic_core/empowerment.hpp"

using namespace intrinsic_core;             // NOLINT(build/namespaces)

class EmpowermentNodePlanar : public rclcpp::Node
{
public:
  EmpowermentNodePlanar()
  : Node("empowerment_node")
  {
    declare_parameter("link1_length", 0.30);
    declare_parameter("link2_length", 0.25);
    declare_parameter("joint_step", 0.12);
    declare_parameter("contact_radius", 0.045);
    declare_parameter("object_mass_factor", 1.0);
    declare_parameter("horizon", 3);
    declare_parameter("n_sequences", 300);
    declare_parameter("outcome_resolution", 0.025);
    declare_parameter("rate_hz", 5.0);
    declare_parameter("joint_names", std::vector<std::string>{"joint1", "joint2"});
    declare_parameter("initial_object_xy", std::vector<double>{0.30, 0.25});
    declare_parameter("frame_id", std::string("base_link"));

    ArmConfig arm_cfg;
    arm_cfg.l1 = get_parameter("link1_length").as_double();
    arm_cfg.l2 = get_parameter("link2_length").as_double();
    arm_cfg.dq = get_parameter("joint_step").as_double();
    arm_cfg.puck_radius = get_parameter("contact_radius").as_double();
    arm_cfg.puck_mass_factor = get_parameter("object_mass_factor").as_double();
    arm_ = std::make_shared<PlanarArm>(arm_cfg);

    EmpowermentConfig cfg;
    cfg.n_steps = static_cast<int>(get_parameter("horizon").as_int());
    cfg.n_sequences = static_cast<int>(get_parameter("n_sequences").as_int());
    cfg.seed = 0;
    auto arm = arm_;
    estimator_ = std::make_shared<EmpowermentEstimator>(
      [arm](const Eigen::VectorXd & s, const std::vector<int> & a) {
        return arm->rollout(s, a);
      },
      9, grid_discretiser(get_parameter("outcome_resolution").as_double()), cfg);

    joint_names_ = get_parameter("joint_names").as_string_array();
    const auto obj = get_parameter("initial_object_xy").as_double_array();
    object_xy_ = Eigen::Vector2d(obj[0], obj[1]);

    sub_joints_ = create_subscription<sensor_msgs::msg::JointState>(
      "joint_states", 10,
      [this](const sensor_msgs::msg::JointState::SharedPtr m) {on_joints(m);});
    sub_object_ = create_subscription<geometry_msgs::msg::PoseStamped>(
      "object_pose", 10,
      [this](const geometry_msgs::msg::PoseStamped::SharedPtr m) {
        object_xy_ = Eigen::Vector2d(m->pose.position.x, m->pose.position.y);
      });

    pub_value_ = create_publisher<std_msgs::msg::Float64>("empowerment/value", 10);
    pub_grad_ =
      create_publisher<geometry_msgs::msg::Vector3>("empowerment/gradient", 10);
    auto latched = rclcpp::QoS(1).reliable().transient_local();
    pub_marker_ = create_publisher<visualization_msgs::msg::MarkerArray>(
      "empowerment/markers", latched);

    const double rate = std::max(get_parameter("rate_hz").as_double(), 1e-3);
    timer_ = create_wall_timer(
      std::chrono::duration<double>(1.0 / rate), [this]() {tick();});

    RCLCPP_INFO(
      get_logger(), "empowerment_node up: horizon=%ld, resolution=%.3f m, rate=%.1f Hz",
      get_parameter("horizon").as_int(),
      get_parameter("outcome_resolution").as_double(), rate);
  }

private:
  void on_joints(const sensor_msgs::msg::JointState::SharedPtr msg)
  {
    Eigen::Vector2d q;
    for (size_t i = 0; i < joint_names_.size() && i < 2; ++i) {
      const auto it = std::find(msg->name.begin(), msg->name.end(), joint_names_[i]);
      if (it == msg->name.end()) {
        RCLCPP_WARN_THROTTLE(
          get_logger(), *get_clock(), 5000, "joint_states missing %s",
          joint_names_[i].c_str());
        return;
      }
      q[static_cast<Eigen::Index>(i)] =
        msg->position[static_cast<size_t>(std::distance(msg->name.begin(), it))];
    }
    q_ = q;
    have_joints_ = true;
  }

  Eigen::VectorXd state_of(const Eigen::Vector2d & q) const
  {
    Eigen::VectorXd s(4);
    s << q[0], q[1], object_xy_[0], object_xy_[1];
    return s;
  }

  void tick()
  {
    if (!have_joints_) {return;}

    const double value = estimator_->estimate(state_of(q_));
    std_msgs::msg::Float64 out;
    out.data = value;
    pub_value_->publish(out);

    const Eigen::Vector2d gradient = finite_difference_gradient();
    geometry_msgs::msg::Vector3 g;
    g.x = gradient[0];
    g.y = gradient[1];
    g.z = 0.0;
    pub_grad_->publish(g);

    publish_markers(value);
  }

  /// Central difference in joint space.
  ///
  /// Four extra empowerment evaluations per cycle. Empowerment has no
  /// analytic gradient -- it is a count -- so this is the only option, and it
  /// is the main reason the node is slow.
  Eigen::Vector2d finite_difference_gradient()
  {
    const double eps = get_parameter("joint_step").as_double();
    Eigen::Vector2d grad = Eigen::Vector2d::Zero();
    for (int i = 0; i < 2; ++i) {
      Eigen::Vector2d up = q_, down = q_;
      up[i] += eps;
      down[i] -= eps;
      const double e_up = estimator_->estimate(state_of(up));
      const double e_dn = estimator_->estimate(state_of(down));
      grad[i] = (e_up - e_dn) / (2.0 * eps);
    }
    return grad;
  }

  void publish_markers(double value)
  {
    const Eigen::Vector2d e = arm_->forward_kinematics(q_[0], q_[1]);

    visualization_msgs::msg::Marker text;
    text.header.frame_id = get_parameter("frame_id").as_string();
    text.header.stamp = now();
    text.ns = "empowerment";
    text.id = 0;
    text.type = visualization_msgs::msg::Marker::TEXT_VIEW_FACING;
    text.action = visualization_msgs::msg::Marker::ADD;
    text.pose.position.x = e[0];
    text.pose.position.y = e[1];
    text.pose.position.z = 0.12;
    text.pose.orientation.w = 1.0;
    text.scale.z = 0.05;
    text.color.a = 1.0;
    text.color.r = text.color.g = text.color.b = 1.0;
    char buf[32];
    std::snprintf(buf, sizeof(buf), "%.2f bits", value);
    text.text = buf;

    visualization_msgs::msg::MarkerArray array;
    array.markers.push_back(text);
    pub_marker_->publish(array);
  }

  std::shared_ptr<PlanarArm> arm_;
  std::shared_ptr<EmpowermentEstimator> estimator_;
  std::vector<std::string> joint_names_;
  Eigen::Vector2d q_{0.0, 0.0};
  Eigen::Vector2d object_xy_{0.30, 0.25};
  bool have_joints_ = false;

  rclcpp::Subscription<sensor_msgs::msg::JointState>::SharedPtr sub_joints_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr sub_object_;
  rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr pub_value_;
  rclcpp::Publisher<geometry_msgs::msg::Vector3>::SharedPtr pub_grad_;
  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr pub_marker_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<EmpowermentNodePlanar>());
  rclcpp::shutdown();
  return 0;
}

File: intrinsic_motivation_ros/intrinsic_motivation_ros/empowerment_node_planar.py
#!/usr/bin/env python3
"""
empowerment_node -- publishes the arm's n-step empowerment, live.

Subscribes:
    /joint_states                sensor_msgs/JointState
    /object_pose                 geometry_msgs/PoseStamped   (the puck)

Publishes:
    /empowerment/value           std_msgs/Float64            bits
    /empowerment/gradient        geometry_msgs/Vector3       ascent direction
                                                             in joint space
    /empowerment/markers         visualization_msgs/MarkerArray

The node holds an analytic forward model of the arm.  Every control cycle it
replays a batch of candidate action sequences through that model, counts how
many distinguishable outcomes result, and publishes the log of that count.

Nothing here knows what the object is for.  The object enters only because
its position is part of the outcome vector, which is why empowerment rises
near it.
"""

from __future__ import annotations

import numpy as np
import rclpy
from geometry_msgs.msg import PoseStamped, Vector3
from rclpy.node import Node
from rclpy.qos import QoSDurabilityPolicy, QoSProfile, QoSReliabilityPolicy
from sensor_msgs.msg import JointState
from std_msgs.msg import Float64
from visualization_msgs.msg import Marker, MarkerArray

from intrinsic_core import (
    ArmConfig,
    EmpowermentConfig,
    EmpowermentEstimator,
    GridDiscretiser,
    PlanarArm,
)


class EmpowermentNode(Node):
    def __init__(self) -> None:
        super().__init__("empowerment_node")

        self._declare_parameters()
        p = self.get_parameter

        self.arm = PlanarArm(
            ArmConfig(
                l1=p("link1_length").value,
                l2=p("link2_length").value,
                dq=p("joint_step").value,
                puck_radius=p("contact_radius").value,
                puck_mass_factor=p("object_mass_factor").value,
            )
        )

        self.estimator = EmpowermentEstimator(
            rollout=self.arm.rollout,
            n_actions=9,
            discretiser=GridDiscretiser(p("outcome_resolution").value),
            config=EmpowermentConfig(
                n_steps=p("horizon").value,
                n_sequences=p("n_sequences").value,
                seed=0,
            ),
        )

        self.joint_names = list(p("joint_names").value)
        self.q = np.zeros(2)
        self.object_xy = np.array(p("initial_object_xy").value, dtype=float)
        self.have_joints = False

        latched = QoSProfile(
            depth=1,
            reliability=QoSReliabilityPolicy.RELIABLE,
            durability=QoSDurabilityPolicy.TRANSIENT_LOCAL,
        )

        self.create_subscription(JointState, "joint_states", self._on_joints, 10)
        self.create_subscription(PoseStamped, "object_pose", self._on_object, 10)

        self.pub_value = self.create_publisher(Float64, "empowerment/value", 10)
        self.pub_grad = self.create_publisher(Vector3, "empowerment/gradient", 10)
        self.pub_marker = self.create_publisher(
            MarkerArray, "empowerment/markers", latched
        )

        period = 1.0 / max(p("rate_hz").value, 1e-3)
        self.create_timer(period, self._tick)

        self.get_logger().info(
            f"empowerment_node up: horizon={p('horizon').value}, "
            f"resolution={p('outcome_resolution').value} m, "
            f"rate={p('rate_hz').value} Hz"
        )

    # -- parameters ---------------------------------------------------------

    def _declare_parameters(self) -> None:
        self.declare_parameter("link1_length", 0.30)
        self.declare_parameter("link2_length", 0.25)
        self.declare_parameter("joint_step", 0.12)
        self.declare_parameter("contact_radius", 0.045)
        self.declare_parameter("object_mass_factor", 1.0)

        self.declare_parameter("horizon", 3)
        self.declare_parameter("n_sequences", 300)
        self.declare_parameter("outcome_resolution", 0.025)

        self.declare_parameter("rate_hz", 5.0)
        self.declare_parameter("joint_names", ["joint1", "joint2"])
        self.declare_parameter("initial_object_xy", [0.30, 0.25])
        self.declare_parameter("frame_id", "base_link")

    # -- callbacks ----------------------------------------------------------

    def _on_joints(self, msg: JointState) -> None:
        lookup = dict(zip(msg.name, msg.position))
        try:
            self.q = np.array([lookup[n] for n in self.joint_names])
            self.have_joints = True
        except KeyError:
            missing = set(self.joint_names) - set(msg.name)
            self.get_logger().warn(
                f"joint_states missing {missing}", throttle_duration_sec=5.0
            )

    def _on_object(self, msg: PoseStamped) -> None:
        self.object_xy = np.array([msg.pose.position.x, msg.pose.position.y])

    # -- main loop ----------------------------------------------------------

    def _state(self, q: np.ndarray) -> np.ndarray:
        return np.array([q[0], q[1], self.object_xy[0], self.object_xy[1]])

    def _tick(self) -> None:
        if not self.have_joints:
            return

        value = self.estimator.estimate(self._state(self.q))
        self.pub_value.publish(Float64(data=float(value)))

        gradient = self._finite_difference_gradient(value)
        self.pub_grad.publish(
            Vector3(x=float(gradient[0]), y=float(gradient[1]), z=0.0)
        )
        self._publish_markers(value)

    def _finite_difference_gradient(self, here: float) -> np.ndarray:
        """Central difference in joint space.

        Four extra empowerment evaluations per cycle.  Empowerment has no
        analytic gradient -- it is a count -- so this is the only option,
        and it is the main reason the node is slow.
        """
        eps = self.get_parameter("joint_step").value
        grad = np.zeros(2)
        for i in range(2):
            up, down = self.q.copy(), self.q.copy()
            up[i] += eps
            down[i] -= eps
            e_up = self.estimator.estimate(self._state(up))
            e_dn = self.estimator.estimate(self._state(down))
            grad[i] = (e_up - e_dn) / (2.0 * eps)
        return grad

    # -- visualisation ------------------------------------------------------

    def _publish_markers(self, value: float) -> None:
        frame = self.get_parameter("frame_id").value
        ex, ey = self.arm.forward_kinematics(self.q[0], self.q[1])

        text = Marker()
        text.header.frame_id = frame
        text.header.stamp = self.get_clock().now().to_msg()
        text.ns = "empowerment"
        text.id = 0
        text.type = Marker.TEXT_VIEW_FACING
        text.action = Marker.ADD
        text.pose.position.x = float(ex)
        text.pose.position.y = float(ey)
        text.pose.position.z = 0.12
        text.pose.orientation.w = 1.0
        text.scale.z = 0.05
        text.color.a = 1.0
        text.color.r = text.color.g = text.color.b = 1.0
        text.text = f"{value:.2f} bits"

        self.pub_marker.publish(MarkerArray(markers=[text]))


def main(args=None) -> None:
    rclpy.init(args=args)
    node = EmpowermentNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == "__main__":
    main()
